File: app-java/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/activities/investigation/evidence/EvidenceSoloFragment.java
package com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.evidence;

import android.annotation.SuppressLint;
import android.media.MediaPlayer;
import android.os.Bundle;
import android.util.Log;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageView;

import com.TritiumGaming.phasmophobiaevidencepicker.R;
import com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.InvestigationActivity;
import com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.evidence.data.runnables.SanityRunnable;

/**
 * EvidenceSoloFragment class
 *
 * @author TritiumGamingStudios
 */
public class EvidenceSoloFragment extends EvidenceFragment {

    @Nullable
    private Thread sanityThread; //Thread that updates the sanity levels

    /**
     * EvidenceSoloFragment default constructor
     */
    public EvidenceSoloFragment() {
        super(R.layout.fragment_evidence);
    }

    @SuppressLint("UseCompatLoadingForDrawables")
    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {

        super.onViewCreated(view, savedInstanceState);

        AppCompatImageView button_reset = view.findViewById(R.id.button_reset);

        button_reset.setOnClickListener(v -> {
            // TODO animate reset arrow
            reset();
        });

        enableUIThread();
    }

    /**
     * reset method
     */
    public void reset() {
        disableUIThread();

        super.softReset();

        requestInvalidateComponents();

        enableUIThread();
    }

    @Override
    public void requestInvalidateComponents() {
        super.requestInvalidateComponents();
    }

    /**
     * enableUIThread method
     */
    public void enableUIThread() {

        if (evidenceViewModel != null && evidenceViewModel.getSanityRunnable() == null) {

            try {
                String appLang = ((InvestigationActivity) requireActivity()).getAppLanguage();
                evidenceViewModel.setSanityRunnable(
                        new SanityRunnable(
                            evidenceViewModel,
                            globalPreferencesViewModel,
                            getHuntWarningAudio(appLang)
                        )
                );
            } catch (IllegalStateException e) {
                e.printStackTrace();
            }

            if (sanityThread == null) {

                if (evidenceViewModel.hasTimerModel()) {
                    evidenceViewModel.getTimerModel().playTimer();
                }

                if (evidenceViewModel.hasSanityRunnable()) {

                    sanityThread = new Thread() {
                        public void run() {
                            while (evidenceViewModel != null && evidenceViewModel.hasSanityModel()
                                    && !evidenceViewModel.getTimerModel().getPaused().getValue()) {
                                try {
                                    update();
                                    tick();
                                } catch (InterruptedException e) {
                                    Log.e("EvidenceFragment",
                                            "(SanityThread) InterruptedException error " +
                                                    "handled.");
                                }
                            }
                        }

                        private void tick() throws InterruptedException {
                            long now = System.nanoTime();
                            long updateTime = System.nanoTime() - now;
                            int TARGET_FPS = 30;
                            long OPTIMAL_TIME = 1000000000 / TARGET_FPS;
                            long wait = (OPTIMAL_TIME - updateTime) / 1000000;

                            if (wait < 0) {
                                wait = 1;
                            }
                            evidenceViewModel.getSanityRunnable().setWait(wait);
                            Thread.sleep(wait);
                        }

                        private void update() {
                            try {
                                requireActivity().runOnUiThread(
                                        evidenceViewModel.getSanityRunnable());
                            } catch (IllegalStateException e) {
                                e.printStackTrace();
                            }
                        }
                    };
                    sanityThread.start();
                }
            }
        }
    }

    /**
     * disableUIThread method
     */
    public void disableUIThread() {
        if (evidenceViewModel != null) {

            if (evidenceViewModel.hasTimerModel()) {
                evidenceViewModel.getTimerModel().pauseTimer();
            }

            SanityRunnable sanityRunnable = evidenceViewModel.getSanityRunnable();

            if (sanityThread != null) {
                sanityThread.interrupt();

                if (sanityRunnable != null) {
                    sanityRunnable.haltMediaPlayer();
                    sanityRunnable.dereferenceViews();
                    evidenceViewModel.setSanityRunnable(null);
                }

                sanityThread = null;
            }

        }

    }

    public MediaPlayer getHuntWarningAudio(@NonNull String appLang) {
        MediaPlayer p = null;
        try {
            p = switch (appLang) {
                case "es" -> MediaPlayer.create(requireContext(), R.raw.huntwarning_es);
                case "fr" -> MediaPlayer.create(requireContext(), R.raw.huntwarning_fr);
                case "de" -> MediaPlayer.create(requireContext(), R.raw.huntwarning_de);
                default -> MediaPlayer.create(requireContext(), R.raw.huntwarning_en);
            };
        } catch (IllegalStateException e) {
            e.printStackTrace();
        }
        return p;
    }

    /**
     * onPause method
     */
    @Override
    public void onPause() {
        disableUIThread();

        if (evidenceViewModel != null && evidenceViewModel.hasSanityModel() &&
            evidenceViewModel.hasSanityRunnable()) {
            evidenceViewModel.getSanityRunnable().haltMediaPlayer();
            evidenceViewModel.getSanityRunnable().dereferenceViews();
        }

        super.onPause();
    }

    /**
     * onResume method
     */
    @Override
    public void onResume() {

        enableUIThread();

        super.onResume();
    }

}

File: app-java/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/activities/investigation/evidence/EvidenceSoloFragment.kt
package com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.evidence

import android.annotation.SuppressLint
import android.media.MediaPlayer
import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.appcompat.widget.AppCompatImageView
import com.TritiumGaming.phasmophobiaevidencepicker.R
import com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.InvestigationActivity
import com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.evidence.runnables.SanityRunnable

/**
 * EvidenceSoloFragment class
 *
 * @author TritiumGamingStudios
 */
class EvidenceSoloFragment : EvidenceFragment(R.layout.fragment_evidence) {

    private var sanityThread: Thread? = null //Thread that updates the sanity levels

    @SuppressLint("UseCompatLoadingForDrawables")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val buttonReset = view.findViewById<AppCompatImageView>(R.id.button_reset)

        buttonReset.setOnClickListener {
            // TODO animate reset arrow
            reset()
        }

        enableUIThread()
    }

    /**
     * reset method
     */
    override fun reset() {
        disableUIThread()

        super.reset()

        requestInvalidateComponents()

        enableUIThread()
    }

    /**
     * enableUIThread method
     */
    private fun enableUIThread() {
        if (evidenceViewModel?.sanityRunnable == null) {
            try {
                val appLang = (requireActivity() as InvestigationActivity).appLanguage
                val runnable = SanityRunnable(evidenceViewModel, globalPreferencesViewModel)
                runnable.huntWarningAudioListener =
                    object : SanityRunnable.HuntWarningAudioListener() {
                        override fun init() { mediaPlayer = getHuntWarningAudio(appLang) }
                        override fun play() {
                            if(globalPreferencesViewModel?.isHuntWarningAudioAllowed == true) {
                                mediaPlayer?.start()
                                evidenceViewModel?.sanityModel?.warnTriggered = true
                            }
                        }
                        override fun stop() { mediaPlayer?.release() }
                    }
                evidenceViewModel?.sanityRunnable = runnable
            } catch (e: IllegalStateException) {
                e.printStackTrace()
            }

            Log.d("SanityThread", if(sanityThread == null) "DNE" else "Exists")
            if (sanityThread == null) {
                Log.d("SanityRunnable", if(evidenceViewModel?.sanityRunnable == null) "DNE" else "Exists")
                if (evidenceViewModel?.hasSanityRunnable() == true) {
                    sanityThread = object : Thread() {
                        override fun run() {
                            while (evidenceViewModel?.timerModel?.paused?.value == false) {
                                try {
                                    update()
                                    tick()
                                } catch (e: InterruptedException) {
                                    Log.e("EvidenceFragment",
                                        "(SanityThread) InterruptedException error handled.")
                                }
                            }
                        }

                        @Throws(InterruptedException::class)
                        private fun tick() {
                            val now = System.nanoTime()
                            val updateTime = System.nanoTime() - now
                            val fpsTarget = 30
                            val timeOptimal = (1000000000 / fpsTarget).toLong()

                            var wait = (timeOptimal - updateTime) / 1000000
                            if (wait < 0) { wait = 1 }

                            evidenceViewModel?.sanityRunnable?.wait = wait

                            sleep(wait)
                        }

                        private fun update() {
                            try {
                                requireActivity().runOnUiThread(evidenceViewModel?.sanityRunnable)
                            } catch (e: IllegalStateException) {
                                e.printStackTrace()
                            }
                        }
                    }
                    sanityThread?.start()
                }
            }
        }
    }

    /**
     * disableUIThread method
     */
    private fun disableUIThread() {
        sanityThread?.interrupt()

        evidenceViewModel?.let { evidenceViewModel ->
            evidenceViewModel.sanityRunnable?.huntWarningAudioListener?.stop()
            evidenceViewModel.sanityRunnable = null
        }
        sanityThread = null
    }

    private fun getHuntWarningAudio(appLang: String): MediaPlayer? {
        var p: MediaPlayer? = null
        try {
            p = when (appLang) {
                "es" -> MediaPlayer.create(requireContext(), R.raw.huntwarning_es)
                "fr" -> MediaPlayer.create(requireContext(), R.raw.huntwarning_fr)
                "de" -> MediaPlayer.create(requireContext(), R.raw.huntwarning_de)
                else -> MediaPlayer.create(requireContext(), R.raw.huntwarning_en)
            }
        } catch (e: IllegalStateException) {
            e.printStackTrace()
        }
        return p
    }

    /**
     * onPause method
     */
    override fun onPause() {
        disableUIThread()

        evidenceViewModel?.sanityRunnable?.huntWarningAudioListener?.stop()

        super.onPause()
    }

    /**
     * onResume method
     */
    override fun onResume() {
        enableUIThread()

        super.onResume()
    }
}